lib/Bindings/Python/circt/dialects/_rtl_ops_ext.py

```diff
@@ -0,0 +1,76 @@
+from mlir.ir import *
+
+class RTLModuleOp:
+  """Specialization for the RTL module op class."""
+
+  def __init__(self,
+               name,
+               input_ports,
+               output_ports,
+               *,
+               body_builder=None,
+               loc=None,
+               ip=None):
+    """
+    Create a RTLModuleOp with the provided `name`, `input_ports`, and
+    `output_ports`.
+    - `name` is a string representing the function name.
+    - `input_ports` is a list of pairs of string names and mlir.ir types.
+    - `output_ports` is a list of pairs of string names and mlir.ir types.
+    - `body_builder` is an optional callback, when provided a new entry block
+      is created and the callback is invoked with the new op as argument within
+      an InsertionPoint context already set for the block. The callback is
+      expected to insert a terminator in the block.
+    """
+    operands = []
+    results = []
+    attributes = {}
+    attributes['sym_name'] = StringAttr.get(str(name))
+
+    input_types = []
+    self._build_ports('arg', attributes, input_ports, input_types)
+
+    output_types = []
+    self._build_ports('result', attributes, output_ports, output_types)
+
+    attributes["type"] = TypeAttr.get(FunctionType.get(
+      inputs=input_types, results=output_types))
+
+    super().__init__(self.build_generic(
+        attributes=attributes, results=results, operands=operands,
```

Comment:
I don't understand how this works... This class doesn't extend any other, so how does `super()` work?

Reply:
Same with all the other member variables. Does MLIR detect that this is supposed to extend rtl.RTLModuleOp and do that? When does this step happen?

Reply:
All sugar classes are extending `OpView` in Python, analogous to how all sugar classes extend `Op` in C++. This is in the generated Python:

```py
@_ods_extend_opview_class(_ods_ext_module)
class RTLModuleOp(_ods_ir.OpView):
  ...
```

The `@_ods_extend_opview_class` takes care of applying this extension (in `_rtl_ops_ext` becomes `_ods_ops_ext` in this snippet), so `super` here is `OpView`.

Definition is here: https://github.com/llvm/llvm-project/blob/main/mlir/lib/Bindings/Python/mlir/dialects/_ods_common.py#L16.

Personal side note, I actually think this kind of metaprogramming is much cleaner in Ruby. Not sure if anyone else would want a Ruby extension though...

Reply:
Sorry, was missing docs: https://reviews.llvm.org/D99387

Reply:
Ruby is much better for meta-programming ime. (but I think this thing is best prioritized on which bindings/styles meet the most people in a given domain and Python is pretty dominant)